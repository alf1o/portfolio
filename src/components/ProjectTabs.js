import React, { Component } from 'react';
import PropTypes from 'prop-types';
import AppBar from '@material-ui/core/AppBar';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import { Link } from 'react-router-dom';
import { withStyles } from '@material-ui/core/styles';

const styles = theme => ({
  tabs: {
    backgroundColor: theme.palette.background.paper
  }
});

/**
  Render the tabs for projects navigation.
*/
// TODO: fix click
class ProjectTabs extends Component {

  static propTypes = {
    classes: PropTypes.object.isRequired,
    projects: PropTypes.array.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      currentTab: 0
    };
    this.changeTab = this.changeTab.bind(this);
  }

  changeTab(_, value) {
    this.setState({ currentTab: value });
  }

  render() {
    const { classes, projects } = this.props;
    const { currentTab } = this.state;
    return (
      <AppBar
        position="static"
      >
        <Tabs
          value={currentTab}
          onChange={this.changeTab}
          scrollable={true}
          indicatorColor="primary"
          textColor="primary"
          className={classes.tabs}
        >
          {projects.map(project => (
            <Tab key={project.name}
              component={Link}
              to={project.name}
              label={
                project.name
              }
            />
          ))}
        </Tabs>
      </AppBar>
    );
  }
}

export { ProjectTabs };
export default withStyles(styles)(ProjectTabs);
