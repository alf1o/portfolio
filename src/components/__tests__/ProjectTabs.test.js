import React from 'react';
import { ProjectTabs } from '../ProjectTabs';
import { createShallow } from '@material-ui/core/test-utils';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';

describe('`ProjectTabs`', () => {
  let props = {};
  let mountedTabs;
  const shallow = createShallow();

  function tabs() {
    if (!mountedTabs) mountedTabs = shallow(<ProjectTabs {...props} />);
    return mountedTabs;
  }

  beforeEach(() => {
    props = {
      classes: {
        tabs: ''
      },
      match: {
        match: {
          url: 'test.com'
        }
      },
      projects: [
        {
          name: 'hello'
        },
        {
          name: 'world'
        }
      ]
    };
    mountedTabs = undefined;
  });

  it('should have a `state.currentTab` property', () => {
    expect(tabs().state().currentTab).toBeDefined();
  });

  it('should have a `changeTab` method', () => {
    expect(ProjectTabs.prototype.changeTab).toBeDefined();
  });

  describe('`changeTab`', () => {
    const changeTabSpy = jest.spyOn(ProjectTabs.prototype, 'changeTab');

    afterEach(() => {
      changeTabSpy.mockClear();
    });

    it('should update `currentTab`', () => {
      const ts = tabs().find(Tabs);
      expect(tabs().state().currentTab).toBe(0);
      ts.simulate('change', null, 3);
      expect(tabs().state().currentTab).toBe(3);
      ts.simulate('change', null, 1);
      expect(tabs().state().currentTab).toBe(1);
    });
  });

  describe('`Tabs`', () => {
    it('should receive a `value` prop which is `changeTab`', () => {
      const ts = tabs().find(Tabs);
      expect(ts.props().value).toBe(tabs().state().currentTab);
    });

    it('should receive an `onChange` prop which is `changeTab`', () => {
      const ts = tabs().find(Tabs);
      expect(ts.props().onChange).toEqual(tabs().instance().changeTab);
    });
  });

  it('should render as many `Tab`s as `props.projects`', () => {
    const len = props.projects.length;
    const ts = tabs().find(Tab);
    expect(ts.length).toBe(len);
  });

  describe('`Tab`', () => {
    it('should receive a `to` prop which is `url` + `project.name`', () => {
      const ts = tabs().find(Tab);
      const url = props.match.match.url;
      ts.forEach((t, i) => {
        expect(t.props().to).toBe(url + '/' + props.projects[i].name);
      });
    });

    it('should receive a `label` prop `project.name`', () => {
      const ts = tabs().find(Tab);
      ts.forEach((t, i) => {
        expect(t.props().label).toBe(props.projects[i].name);
      });
    });
  });

  it('should match snapshot', () => {
    expect(tabs()).toMatchSnapshot();
  });
});
