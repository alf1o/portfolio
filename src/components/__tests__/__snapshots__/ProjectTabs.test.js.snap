// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`ProjectTabs\` should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProjectTabs
    classes={
        Object {
            "tabs": "",
          }
    }
    match={
        Object {
            "match": Object {
              "url": "test.com",
            },
          }
    }
    projects={
        Array [
            Object {
              "name": "hello",
            },
            Object {
              "name": "world",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <WithStyles(Tabs)
        className=""
        indicatorColor="primary"
        onChange={[Function]}
        scrollable={true}
        textColor="primary"
        value={0}
>
        <WithStyles(Tab)
                component={[Function]}
                label="hello"
                to="test.com/hello"
        />
        <WithStyles(Tab)
                component={[Function]}
                label="world"
                to="test.com/world"
        />
</WithStyles(Tabs)>,
      "position": "static",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Tab)
            component={[Function]}
            label="hello"
            to="test.com/hello"
/>,
          <WithStyles(Tab)
            component={[Function]}
            label="world"
            to="test.com/world"
/>,
        ],
        "className": "",
        "indicatorColor": "primary",
        "onChange": [Function],
        "scrollable": true,
        "textColor": "primary",
        "value": 0,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "hello",
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "label": "hello",
            "to": "test.com/hello",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "world",
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "label": "world",
            "to": "test.com/world",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <WithStyles(Tabs)
          className=""
          indicatorColor="primary"
          onChange={[Function]}
          scrollable={true}
          textColor="primary"
          value={0}
>
          <WithStyles(Tab)
                    component={[Function]}
                    label="hello"
                    to="test.com/hello"
          />
          <WithStyles(Tab)
                    component={[Function]}
                    label="world"
                    to="test.com/world"
          />
</WithStyles(Tabs)>,
        "position": "static",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(Tab)
              component={[Function]}
              label="hello"
              to="test.com/hello"
/>,
            <WithStyles(Tab)
              component={[Function]}
              label="world"
              to="test.com/world"
/>,
          ],
          "className": "",
          "indicatorColor": "primary",
          "onChange": [Function],
          "scrollable": true,
          "textColor": "primary",
          "value": 0,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "hello",
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "label": "hello",
              "to": "test.com/hello",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "world",
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "label": "world",
              "to": "test.com/world",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
